File: packages/node/src/agent-hooks.js
import { createHash } from "node:crypto";
import {
  mkdirSync,
  readFileSync,
  readSync,
  renameSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { dirname, join } from "node:path";

import { newObservation } from "./observation.js";
import { NDJSONStore } from "./storage.js";

export const MAX_HOOK_INPUT_BYTES = 8 * 1024 * 1024;
const CLIENTS = {
  "claude-code": { provider: "anthropic", keepModel: true },
  codex: { provider: "openai", keepModel: false },
};
const MODEL = /^[A-Za-z0-9][A-Za-z0-9._:/+\-]{0,255}$/;
const FAILURE_TYPES = new Set([
  "rate_limit",
  "overloaded",
  "authentication_failed",
  "oauth_org_not_allowed",
  "billing_error",
  "invalid_request",
  "model_not_found",
  "server_error",
  "max_output_tokens",
  "timeout",
  "unknown",
]);

function enabled() {
  return !["1", "true", "yes", "on"].includes(
    String(process.env.LLMWHO_DISABLED ?? "").trim().toLowerCase(),
  );
}

function safeModel(value) {
  return typeof value === "string" && MODEL.test(value) ? value : undefined;
}

function statePath(store, client, sessionId) {
  if (typeof sessionId !== "string" || !sessionId) return undefined;
  const digest = createHash("sha256").update(sessionId, "utf8").digest("hex");
  return join(dirname(store.path), ".hook-state", client, `${digest}.json`);
}

function readState(path) {
  if (!path) return {};
  let value;
  try {
    value = JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return {};
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const state = {};
  const model = safeModel(value.model);
  if (model) state.model = model;
  if (Number.isFinite(value.started_at_ms) && value.started_at_ms >= 0) {
    state.started_at_ms = value.started_at_ms;
  }
  if (
    Number.isSafeInteger(value.input_bytes)
    && value.input_bytes >= 0
    && value.input_bytes <= MAX_HOOK_INPUT_BYTES
  ) {
    state.input_bytes = value.input_bytes;
  }
  return state;
}

function deleteState(path) {
  if (!path) return;
  try {
    unlinkSync(path);
  } catch (error) {
    if (error.code !== "ENOENT") return;
  }
}

function writeState(path, state) {
  if (!path) return;
  const safe = {};
  const model = safeModel(state.model);
  if (model) safe.model = model;
  if (Number.isFinite(state.started_at_ms) && state.started_at_ms >= 0) {
    safe.started_at_ms = state.started_at_ms;
  }
  if (
    Number.isSafeInteger(state.input_bytes)
    && state.input_bytes >= 0
    && state.input_bytes <= MAX_HOOK_INPUT_BYTES
  ) {
    safe.input_bytes = state.input_bytes;
  }
  if (!Object.keys(safe).length) {
    deleteState(path);
    return;
  }
  mkdirSync(dirname(path), { recursive: true, mode: 0o700 });
  const temporary = `${path}.${process.pid}.tmp`;
  writeFileSync(temporary, JSON.stringify(safe), { encoding: "utf8", mode: 0o600 });
  renameSync(temporary, path);
}

function durationMs(state, nowMs) {
  return Number.isFinite(state.started_at_ms)
    ? Math.max(0, nowMs - state.started_at_ms)
    : 0;
}

function failureType(value) {
  return typeof value === "string" && FAILURE_TYPES.has(value) ? value : "unknown";
}

export function observeHookEvent(client, payload, {
  store,
  expectedEvent,
  nowMs = Date.now(),
}) {
  const config = CLIENTS[client];
  if (!config) throw new Error("unsupported hook client");
  const eventName = payload?.hook_event_name;
  if (typeof eventName !== "string") return undefined;
  if (expectedEvent && eventName !== expectedEvent) return undefined;

  const currentMs = Math.max(0, nowMs);
  const path = statePath(store, client, payload.session_id);
  const state = readState(path);
  const model = safeModel(payload.model);
  if (model) state.model = model;

  if (eventName === "SessionStart") {
    writeState(path, state);
    return undefined;
  }
  if (eventName === "UserPromptSubmit") {
    state.started_at_ms = currentMs;
    if (typeof payload.prompt === "string") {
      state.input_bytes = Buffer.byteLength(payload.prompt);
    } else {
      delete state.input_bytes;
    }
    writeState(path, state);
    return undefined;
  }
  if (eventName === "SessionEnd") {
    deleteState(path);
    return undefined;
  }
  if (!["Stop", "StopFailure"].includes(eventName)) return undefined;

  const claimedModel = model ?? safeModel(state.model);
  const request = { operation: "agent.turn" };
  if (Number.isSafeInteger(state.input_bytes)) request.input_bytes = state.input_bytes;
  if (claimedModel) request.claimed_model = claimedModel;

  let response;
  let outcome = "success";
  let errorType;
  if (eventName === "Stop") {
    if (typeof payload.last_assistant_message === "string") {
      response = { output_bytes: Buffer.byteLength(payload.last_assistant_message) };
    }
  } else {
    errorType = failureType(payload.error);
    outcome = errorType === "timeout" ? "timeout" : "http_error";
  }

  const event = newObservation({
    url: `hook://${client}/${eventName.toLowerCase()}`,
    durationMs: durationMs(state, currentMs),
    outcome,
    errorType,
    provider: config.provider,
    claimedModel,
    request,
    response,
  });
  store.append(event);

  delete state.started_at_ms;
  delete state.input_bytes;
  if (config.keepModel) writeState(path, state);
  else deleteState(path);
  return event;
}

function readStdin() {
  const chunks = [];
  const buffer = Buffer.allocUnsafe(64 * 1024);
  let total = 0;
  let tooLarge = false;
  while (true) {
    const size = readSync(0, buffer, 0, buffer.length, null);
    if (size === 0) break;
    total += size;
    if (total > MAX_HOOK_INPUT_BYTES) {
      tooLarge = true;
      continue;
    }
    chunks.push(Buffer.from(buffer.subarray(0, size)));
  }
  return tooLarge ? undefined : Buffer.concat(chunks);
}

function writeProtocolOutput(client, eventName) {
  if (client === "codex" && ["Stop", "SubagentStop"].includes(eventName)) {
    process.stdout.write("{}\n");
  }
}

export function runHookCli({ client, expectedEvent, storagePath }) {
  let actualEvent;
  try {
    const raw = readStdin();
    if (raw) {
      const payload = JSON.parse(raw.toString("utf8"));
      if (payload && typeof payload === "object" && !Array.isArray(payload)) {
        actualEvent = payload.hook_event_name;
        if (enabled()) {
          observeHookEvent(client, payload, {
            store: new NDJSONStore(storagePath ?? process.env.LLMWHO_STORAGE),
            expectedEvent,
          });
        }
      }
    }
  } catch {
    // Hooks must remain fail-open and must never replace agent control flow.
  }
  writeProtocolOutput(client, expectedEvent ?? actualEvent);
  return 0;
}

File: packages/python/src/llmwho/agent_hooks.py
"""Fail-open adapters for Claude Code and Codex lifecycle hooks."""

import hashlib
import json
import math
import os
from pathlib import Path
import re
import sys
import time
from typing import Any
from collections.abc import Mapping

from .observation import new_observation
from .storage import JSONLStore


MAX_HOOK_INPUT_BYTES = 8 * 1024 * 1024
CLIENTS = {
    "claude-code": {"provider": "anthropic", "keep_model": True},
    "codex": {"provider": "openai", "keep_model": False},
}
_MODEL = re.compile(r"[A-Za-z0-9][A-Za-z0-9._:/+\-]{0,255}\Z")
_FAILURE_TYPES = frozenset(
    {
        "rate_limit",
        "overloaded",
        "authentication_failed",
        "oauth_org_not_allowed",
        "billing_error",
        "invalid_request",
        "model_not_found",
        "server_error",
        "max_output_tokens",
        "timeout",
        "unknown",
    }
)


def _enabled() -> bool:
    return os.environ.get("LLMWHO_DISABLED", "").strip().lower() not in {
        "1",
        "true",
        "yes",
        "on",
    }


def _safe_model(value: Any) -> str | None:
    return value if isinstance(value, str) and _MODEL.fullmatch(value) else None


def _state_path(store: JSONLStore, client: str, session_id: Any) -> Path | None:
    if not isinstance(session_id, str) or not session_id:
        return None
    digest = hashlib.sha256(session_id.encode("utf-8")).hexdigest()
    return store.path.parent / ".hook-state" / client / f"{digest}.json"


def _read_state(path: Path | None) -> dict[str, Any]:
    if path is None:
        return {}
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return {}
    if not isinstance(value, dict):
        return {}
    state: dict[str, Any] = {}
    model = _safe_model(value.get("model"))
    if model:
        state["model"] = model
    started = value.get("started_at_ms")
    if (
        type(started) in {int, float}
        and math.isfinite(started)
        and started >= 0
    ):
        state["started_at_ms"] = float(started)
    input_bytes = value.get("input_bytes")
    if (
        type(input_bytes) is int
        and 0 <= input_bytes <= MAX_HOOK_INPUT_BYTES
    ):
        state["input_bytes"] = input_bytes
    return state


def _delete_state(path: Path | None) -> None:
    if path is None:
        return
    try:
        path.unlink(missing_ok=True)
    except Exception:
        return


def _write_state(path: Path | None, state: Mapping[str, Any]) -> None:
    if path is None:
        return
    safe: dict[str, Any] = {}
    model = _safe_model(state.get("model"))
    if model:
        safe["model"] = model
    started = state.get("started_at_ms")
    if (
        type(started) in {int, float}
        and math.isfinite(started)
        and started >= 0
    ):
        safe["started_at_ms"] = float(started)
    input_bytes = state.get("input_bytes")
    if (
        type(input_bytes) is int
        and 0 <= input_bytes <= MAX_HOOK_INPUT_BYTES
    ):
        safe["input_bytes"] = input_bytes
    if not safe:
        _delete_state(path)
        return
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    temporary = path.with_name(f".{path.name}.{os.getpid()}.tmp")
    descriptor = os.open(
        temporary,
        os.O_CREAT | os.O_TRUNC | os.O_WRONLY,
        0o600,
    )
    try:
        os.write(
            descriptor,
            json.dumps(safe, separators=(",", ":")).encode("utf-8"),
        )
    finally:
        os.close(descriptor)
    os.replace(temporary, path)


def _duration_ms(state: Mapping[str, Any], now_ms: float) -> float:
    started = state.get("started_at_ms")
    if type(started) not in {int, float}:
        return 0.0
    return max(0.0, now_ms - float(started))


def _failure_type(value: Any) -> str:
    return value if isinstance(value, str) and value in _FAILURE_TYPES else "unknown"


def observe_hook_event(
    client: str,
    payload: Mapping[str, Any],
    *,
    store: JSONLStore,
    expected_event: str | None = None,
    now_ms: float | None = None,
) -> dict[str, Any] | None:
    """Update safe hook state and append one turn observation when applicable."""

    config = CLIENTS.get(client)
    if config is None:
        raise ValueError("unsupported hook client")
    event_name = payload.get("hook_event_name")
    if not isinstance(event_name, str):
        return None
    if expected_event and event_name != expected_event:
        return None

    current_ms = time.time() * 1000 if now_ms is None else max(0.0, now_ms)
    path = _state_path(store, client, payload.get("session_id"))
    state = _read_state(path)
    model = _safe_model(payload.get("model"))
    if model:
        state["model"] = model

    if event_name == "SessionStart":
        _write_state(path, state)
        return None
    if event_name == "UserPromptSubmit":
        state["started_at_ms"] = current_ms
        prompt = payload.get("prompt")
        if isinstance(prompt, str):
            state["input_bytes"] = len(prompt.encode("utf-8"))
        else:
            state.pop("input_bytes", None)
        _write_state(path, state)
        return None
    if event_name == "SessionEnd":
        _delete_state(path)
        return None
    if event_name not in {"Stop", "StopFailure"}:
        return None

    duration_ms = _duration_ms(state, current_ms)
    requested_model = model or _safe_model(state.get("model"))
    request: dict[str, Any] = {"operation": "agent.turn"}
    input_bytes = state.get("input_bytes")
    if type(input_bytes) is int:
        request["input_bytes"] = input_bytes
    if requested_model:
        request["requested_model"] = requested_model

    response = None
    outcome = "success"
    error_type = None
    if event_name == "Stop":
        assistant_message = payload.get("last_assistant_message")
        if isinstance(assistant_message, str):
            response = {"output_bytes": len(assistant_message.encode("utf-8"))}
    else:
        error_type = _failure_type(payload.get("error"))
        outcome = "timeout" if error_type == "timeout" else "http_error"

    event = new_observation(
        url=f"hook://{client}/{event_name.lower()}",
        duration_ms=duration_ms,
        outcome=outcome,
        provider=str(config["provider"]),
        requested_model=requested_model,
        request=request,
        response=response,
        error_type=error_type,
    )
    store.append(event)

    state.pop("started_at_ms", None)
    state.pop("input_bytes", None)
    if config["keep_model"]:
        _write_state(path, state)
    else:
        _delete_state(path)
    return event


def _read_stdin() -> bytes | None:
    stream = getattr(sys.stdin, "buffer", sys.stdin)
    value = stream.read(MAX_HOOK_INPUT_BYTES + 1)
    raw = value.encode("utf-8") if isinstance(value, str) else bytes(value)
    if len(raw) <= MAX_HOOK_INPUT_BYTES:
        return raw
    while stream.read(64 * 1024):
        pass
    return None


def _write_protocol_output(client: str, event_name: str | None) -> None:
    if client == "codex" and event_name in {"Stop", "SubagentStop"}:
        sys.stdout.write("{}\n")


def run_hook_cli(
    client: str,
    *,
    expected_event: str | None = None,
    storage_path: str | None = None,
) -> int:
    """Read one hook payload. Telemetry failure never changes hook control flow."""

    actual_event = None
    try:
        raw = _read_stdin()
        if raw is not None:
            value = json.loads(raw.decode("utf-8"))
            if isinstance(value, dict):
                actual_event = value.get("hook_event_name")
                if _enabled():
                    configured_path = storage_path or os.environ.get("LLMWHO_STORAGE")
                    observe_hook_event(
                        client,
                        value,
                        store=JSONLStore(configured_path),
                        expected_event=expected_event,
                    )
    except Exception:
        pass
    _write_protocol_output(client, expected_event or actual_event)
    return 0
